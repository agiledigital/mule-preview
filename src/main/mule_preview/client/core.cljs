(ns mule-preview.client.core
  "The entry point to the web application"
  (:require
   [reagent.core :as r]
   [mule-preview.client.views.preview :refer [start-preview start-preview-url]]
   [mule-preview.client.views.diff :refer [start-diff start-diff-url]]))

(defn view [root-component]
  [:div {:class "mp root-component"} @root-component])

(defn mount-root [element root-component]
  (r/render [(partial view root-component)] element))

(defn ^:export mount-url-diff-on-element [element file-a file-b content-root]
  (let [root-component (r/atom [:div])]
    (start-diff-url file-a
                    file-b
                    root-component
                    content-root)
    (mount-root element root-component)))

(defn ^:export mount-url-preview-on-element [element file content-root]
  (let [root-component (r/atom [:div])]
    (start-preview-url
     file
     root-component
     content-root)
    (mount-root element root-component)))

(defn ^:export mount-diff-on-element [element content-a content-b content-root]
  (let [root-component (r/atom [:div])]
    (start-diff content-a
                content-b
                root-component
                content-root)
    (mount-root element root-component)))

(defn ^:export mount-preview-on-element [element content content-root]
  (let [root-component (r/atom [:div])]
    (start-preview
     content
     root-component
     content-root)
    (mount-root element root-component)))

(defn ^:dev/after-load init! []
  (mount-url-diff-on-element
   (.getElementById js/document "app")
   "/example_xml/nice-example.xml"
   "/example_xml/nice-example-diff.xml"
   "."))
