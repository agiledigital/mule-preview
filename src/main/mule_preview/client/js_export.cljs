(ns mule-preview.client.js-export
  "Entry point for dependant Javascript projects.
   Exports the function names with camel case to fit Javascript conventions"
  (:require
   [mule-preview.client.core :refer [mount-url-diff-on-element
                                     mount-url-preview-on-element
                                     mount-diff-on-element
                                     mount-preview-on-element]]))

(def ^:export mountUrlDiffOnElement mount-url-diff-on-element)
(def ^:export mountUrlPreviewOnElement mount-url-preview-on-element)
(def ^:export mountDiffOnElement mount-diff-on-element)
(def ^:export mountPreviewOnElement mount-preview-on-element)

