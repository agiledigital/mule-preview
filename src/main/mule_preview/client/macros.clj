(ns mule-preview.client.macros
  "Utility functions that transform the code at compile time")

; https://github.com/reagent-project/reagent/wiki/Beware-Event-Handlers-Returning-False
(defmacro handler-fn
  ([& body]
   `(fn [~'event] ~@body nil)))  ;; force return nil