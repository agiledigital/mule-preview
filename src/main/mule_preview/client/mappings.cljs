(ns mule-preview.client.mappings
  "Mappings that declare how to render Mule components according to their tag name."
  (:require-macros [mule-preview.client.macros :as m]))

(def root-container
  "The root element that every Mule file should have. There should only be one of these per file."
  "mule")

(def error-handler-container-list
  "Elements that are a container, but can also have error handling"
  #{"flow" "transactional" "munit:test"})

(def vertical-container-list
  "Containers that lay out their children vertically"
  #{"scatter-gather" "choice" "composite-source" "round-robin"})

(def horizontal-container-list
  "Containers that lay out their children horizontally"
  #{"flow" "sub-flow" "async" "batch:job"
    "batch:step" "batch:commit" "ee:cache" "foreach"
    "enricher" "poll" "request-reply"
    "until-successful" "when" "otherwise" "processor-chain"
    "munit:test" "munit:before-test" "munit:after-test"
    "munit:after-suite" "munit:before-suite"})

(def error-handler-component-list
  "Elements that can be in the error section of an error handler container"
  #{"catch-exception-strategy" "choice-exception-strategy"
    "exception-strategy" "rollback-exception-strategy"})

(def munit-container-list
  "MUnit test flows"
  #{"munit:test" "munit:before-test" "munit:after-test"
    "munit:after-suite" "munit:before-suite"})
