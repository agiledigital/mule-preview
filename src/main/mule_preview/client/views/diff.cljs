(ns mule-preview.client.views.diff
  "Renders a page that shows a diff between two Mule XML files"
  (:require
   [mule-preview.client.react :refer [mast->react]]
   [mule-preview.client.mast :refer [xml->mast]]
   [mule-preview.client.transformers.apply_patch :refer [augment-mast-with-diff]]
   [mule-preview.client.diff-algorithms.diff-dom :refer [diff]]
   [mule-preview.client.views.common :refer [fetch-mappings]]
   [reagent.core :as r]
   [cljs.core.async :refer [<!]]
   [cljs-http.client :as http]
   [tubax.core :refer [xml->clj]])
  (:require-macros [cljs.core.async.macros :refer [go]]))

(def empty-file "<?xml version=\"1.0\" encoding=\"UTF-8\"?><mule></mule>")

(defn calculate-diff [content-a content-b content-root]
  (let [parsed-xml-a (xml->clj (or content-a empty-file))
        parsed-xml-b (xml->clj (or content-b empty-file))
        mast-a (xml->mast parsed-xml-a)
        mast-b (xml->mast parsed-xml-b)
        diff-output (diff mast-a mast-b)]
    (augment-mast-with-diff mast-a diff-output)))

(defn handle-xml-fetch-success [mappings response-a response-b root-component content-root]
  (let [augmented-mast (calculate-diff response-a response-b content-root)
        transformed-components (mast->react augmented-mast mappings content-root)]
    (reset! root-component transformed-components)))

(defn start-diff-url [url-a url-b root-component content-root]
  (go (let [mappings-response (<! (fetch-mappings content-root))
            mule-xml-response-a (when url-a (<! (http/get url-a)))
            mule-xml-response-b (when url-b (<! (http/get url-b)))]
        (handle-xml-fetch-success
         (:body mappings-response)
         (when mule-xml-response-a  (str (:body mule-xml-response-a)))
         (when mule-xml-response-b  (str (:body mule-xml-response-b)))
         root-component
         content-root))))

(defn start-diff [content-a content-b root-component content-root]
  (go (let [mappings-response (<! (fetch-mappings content-root))]
        (handle-xml-fetch-success
         (:body mappings-response)
         content-a
         content-b
         root-component
         content-root))))