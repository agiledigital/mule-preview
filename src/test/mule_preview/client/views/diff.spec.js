import { readFileSync } from "fs";
import { join } from "path";
import { calculate_diff } from "../../../../../dist/mule_preview.client.views.diff";
import { cljToJs } from "../../../../../dist/mule_preview.client.test_utils";

describe("when diffing two XML files", () => {
  describe("when file B adds elements with nested content", () => {
    it("augments the MAST correctly", () => {
      const xmlFileA = readFileSync(
        join(__dirname, "__fixtures__/nested-add-a.xml")
      );
      const xmlFileB = readFileSync(
        join(__dirname, "__fixtures__/nested-add-b.xml")
      );
      const output = cljToJs(calculate_diff(xmlFileA, xmlFileB, "."));
      expect(output).toMatchSnapshot();
    });
  });

  describe("when diffing MUnit tests", () => {
    it("augments the MAST correctly", () => {
      const xmlFileA = readFileSync(
        join(__dirname, "__fixtures__/munit-a.xml")
      );
      const xmlFileB = readFileSync(
        join(__dirname, "__fixtures__/munit-b.xml")
      );
      const output = cljToJs(calculate_diff(xmlFileA, xmlFileB, "."));
      expect(output).toMatchSnapshot();
    });
  });

  describe("when doing a diff with a non existant initial state", () => {
    it("augments the MAST correctly", () => {
      const xmlFile = readFileSync(join(__dirname, "__fixtures__/munit-a.xml"));
      const output = cljToJs(calculate_diff(undefined, xmlFile, "."));
      expect(output).toMatchSnapshot();
    });
  });

  describe("when doing a diff with a non existant final state", () => {
    it("augments the MAST correctly", () => {
      const xmlFile = readFileSync(join(__dirname, "__fixtures__/munit-b.xml"));
      const output = cljToJs(calculate_diff(xmlFile, undefined, "."));
      expect(output).toMatchSnapshot();
    });
  });
});
