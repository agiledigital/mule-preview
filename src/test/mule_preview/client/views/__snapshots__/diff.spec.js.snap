// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when diffing two XML files when diffing MUnit tests augments the MAST correctly 1`] = `
Object {
  "attributes": Object {
    "xmlns": "http://www.mulesoft.org/schema/mule/core",
    "xmlns:doc": "http://www.mulesoft.org/schema/mule/documentation",
    "xmlns:mock": "http://www.mulesoft.org/schema/mule/mock",
    "xmlns:munit": "http://www.mulesoft.org/schema/mule/munit",
    "xmlns:spring": "http://www.springframework.org/schema/beans",
    "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
    "xsi:schemaLocation": "http://www.mulesoft.org/schema/mule/mock http://www.mulesoft.org/schema/mule/mock/current/mule-mock.xsd
http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd
http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd",
  },
  "content": Array [
    Object {
      "attributes": Object {
        "doc:name": "MUnit configuration",
        "name": "munit",
      },
      "content-hash": -2017569654,
      "description": "MUnit configuration",
      "hash": -910446770,
      "labels": Array [],
      "location": Object {
        "column": 63,
        "line": 13,
      },
      "tag-name": "munit:config",
      "type": "component",
    },
    Object {
      "attributes": Object {
        "description": "Ater suite actions",
        "name": "new-test-suiteAfter_Suite",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "After suite",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "actualValue": "#[]",
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert Equals",
              "doc:name": "Assert Equals",
              "expectedValue": "#[]",
              "hash": 1540583787,
              "labels": Array [
                "added",
              ],
              "location": Object {
                "column": 96,
                "line": 15,
              },
              "tag-name": "munit:assert-on-equals",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "condition": "#[]",
                "doc:name": "Assert False",
              },
              "content-hash": -2017569654,
              "description": "Assert False",
              "hash": -964144491,
              "labels": Array [],
              "location": Object {
                "column": 69,
                "line": 15,
              },
              "tag-name": "munit:assert-false",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteAfter_Suite",
      "labels": Array [
        "munit",
        "vertical",
      ],
      "location": Object {
        "column": 89,
        "line": 14,
      },
      "tag-name": "munit:after-suite",
      "type": "munit-container",
    },
    Object {
      "attributes": Object {
        "description": "After tests actions",
        "name": "new-test-suiteAfter_Test",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "After test",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "actualValue": "#[]",
                "doc:name": "Assert Not Equals",
                "expectedValue": "#[]",
              },
              "content-hash": -2017569654,
              "description": "Assert Not Equals",
              "hash": 1131396067,
              "labels": Array [],
              "location": Object {
                "column": 99,
                "line": 18,
              },
              "tag-name": "munit:assert-not-same",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Assert Not Null Payload",
              },
              "content-hash": -2017569654,
              "description": "Assert Not Null Payload",
              "hash": -802864801,
              "labels": Array [],
              "location": Object {
                "column": 67,
                "line": 19,
              },
              "tag-name": "munit:assert-not-null",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteAfter_Test",
      "labels": Array [
        "munit",
        "vertical",
        "removed",
      ],
      "location": Object {
        "column": 88,
        "line": 17,
      },
      "tag-name": "munit:after-test",
      "type": "munit-container",
    },
    Object {
      "content": Array [
        Object {
          "content": null,
          "description": "Before suite",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert Null Payload",
              "doc:name": "Assert Null Payload",
              "hash": 358321917,
              "labels": Array [],
              "location": Object {
                "column": 59,
                "line": 19,
              },
              "tag-name": "munit:assert-null",
              "type": "component",
            },
            Object {
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert Payload",
              "doc:name": "Assert Payload",
              "expectedValue": "#[]",
              "hash": 1170098794,
              "labels": Array [],
              "location": Object {
                "column": 84,
                "line": 20,
              },
              "tag-name": "munit:assert-payload-equals",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteBefore_Suite",
      "labels": Array [
        "munit",
        "vertical",
        "added",
      ],
      "location": Object {
        "column": 83,
        "line": 18,
      },
      "name": "new-test-suiteBefore_Suite",
      "tag-name": "munit:before-suite",
      "type": "munit-container",
    },
    Object {
      "attributes": Object {
        "description": "Before tests actions",
        "name": "new-test-suiteBefore_Test",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "Before test",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "condition": "#[]",
                "doc:name": "Assert True",
              },
              "content-hash": -2017569654,
              "description": "Assert True",
              "hash": 1099203488,
              "labels": Array [],
              "location": Object {
                "column": 67,
                "line": 22,
              },
              "tag-name": "munit:assert-true",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "assertion-ref": "",
                "doc:name": "Custom Assertion",
              },
              "content-hash": -2017569654,
              "description": "Custom Assertion",
              "hash": 1823118484,
              "labels": Array [],
              "location": Object {
                "column": 72,
                "line": 23,
              },
              "tag-name": "munit:run-custom",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteBefore_Test",
      "labels": Array [
        "munit",
        "vertical",
      ],
      "location": Object {
        "column": 91,
        "line": 21,
      },
      "tag-name": "munit:before-test",
      "type": "munit-container",
    },
    Object {
      "attributes": Object {
        "description": "MUnit Test",
        "name": "new-test-suiteTest",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "Setup",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "doc:name": "Set Message",
                "payload": "#[]",
              },
              "content-hash": -2017569654,
              "description": "Set Message",
              "hash": -339828259,
              "labels": Array [],
              "location": Object {
                "column": 57,
                "line": 26,
              },
              "tag-name": "munit:set",
              "type": "component",
            },
            Object {
              "content": null,
              "content-hash": -2017569654,
              "description": "Fail",
              "doc:name": "Fail",
              "hash": 369682545,
              "labels": Array [
                "added",
              ],
              "location": Object {
                "column": 37,
                "line": 28,
              },
              "tag-name": "munit:fail",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Set Null Payload",
              },
              "content-hash": -2017569654,
              "description": "Set Null Payload",
              "hash": -852187714,
              "labels": Array [],
              "location": Object {
                "column": 61,
                "line": 27,
              },
              "tag-name": "munit:set-null-payload",
              "type": "component",
            },
          ],
          "description": "Test",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteTest",
      "labels": Array [
        "munit",
        "vertical",
      ],
      "location": Object {
        "column": 67,
        "line": 25,
      },
      "tag-name": "munit:test",
      "type": "munit-container",
    },
  ],
  "description": null,
  "hash": -2009730132,
  "labels": Array [
    "vertical",
    "root",
  ],
  "location": Object {
    "column": 108,
    "line": 12,
  },
  "tag-name": "mule",
  "type": "container",
}
`;

exports[`when diffing two XML files when doing a diff with a non existant final state augments the MAST correctly 1`] = `
Object {
  "attributes": Object {
    "xmlns": "http://www.mulesoft.org/schema/mule/core",
    "xmlns:doc": "http://www.mulesoft.org/schema/mule/documentation",
    "xmlns:mock": "http://www.mulesoft.org/schema/mule/mock",
    "xmlns:munit": "http://www.mulesoft.org/schema/mule/munit",
    "xmlns:spring": "http://www.springframework.org/schema/beans",
    "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
    "xsi:schemaLocation": "http://www.mulesoft.org/schema/mule/mock http://www.mulesoft.org/schema/mule/mock/current/mule-mock.xsd
http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd
http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd",
  },
  "content": Array [
    Object {
      "attributes": Object {
        "doc:name": "MUnit configuration",
        "name": "munit",
      },
      "content-hash": -2017569654,
      "description": "MUnit configuration",
      "hash": -910446770,
      "labels": Array [
        "removed",
      ],
      "location": Object {
        "column": 63,
        "line": 13,
      },
      "tag-name": "munit:config",
      "type": "component",
    },
    Object {
      "attributes": Object {
        "description": "Ater suite actions",
        "name": "new-test-suiteAfter_Suite",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "After suite",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "actualValue": "#[]",
                "doc:name": "Assert Equals",
                "expectedValue": "#[]",
              },
              "content-hash": -2017569654,
              "description": "Assert Equals",
              "hash": 1540583787,
              "labels": Array [],
              "location": Object {
                "column": 96,
                "line": 15,
              },
              "tag-name": "munit:assert-on-equals",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "condition": "#[]",
                "doc:name": "Assert False",
              },
              "content-hash": -2017569654,
              "description": "Assert False",
              "hash": -964144491,
              "labels": Array [],
              "location": Object {
                "column": 69,
                "line": 16,
              },
              "tag-name": "munit:assert-false",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteAfter_Suite",
      "labels": Array [
        "munit",
        "vertical",
        "removed",
      ],
      "location": Object {
        "column": 89,
        "line": 14,
      },
      "tag-name": "munit:after-suite",
      "type": "munit-container",
    },
    Object {
      "attributes": Object {
        "description": "MUnit Test",
        "name": "new-test-suiteBefore_Suite",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "Before suite",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "doc:name": "Assert Null Payload",
              },
              "content-hash": -2017569654,
              "description": "Assert Null Payload",
              "hash": 358321917,
              "labels": Array [],
              "location": Object {
                "column": 59,
                "line": 19,
              },
              "tag-name": "munit:assert-null",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Assert Payload",
                "expectedValue": "#[]",
              },
              "content-hash": -2017569654,
              "description": "Assert Payload",
              "hash": 1170098794,
              "labels": Array [],
              "location": Object {
                "column": 84,
                "line": 20,
              },
              "tag-name": "munit:assert-payload-equals",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteBefore_Suite",
      "labels": Array [
        "munit",
        "vertical",
        "removed",
      ],
      "location": Object {
        "column": 83,
        "line": 18,
      },
      "tag-name": "munit:before-suite",
      "type": "munit-container",
    },
    Object {
      "attributes": Object {
        "description": "Before tests actions",
        "name": "new-test-suiteBefore_Test",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "Before test",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "condition": "#[]",
                "doc:name": "Assert True",
              },
              "content-hash": -2017569654,
              "description": "Assert True",
              "hash": 1099203488,
              "labels": Array [],
              "location": Object {
                "column": 67,
                "line": 23,
              },
              "tag-name": "munit:assert-true",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "assertion-ref": "",
                "doc:name": "Custom Assertion",
              },
              "content-hash": -2017569654,
              "description": "Custom Assertion",
              "hash": 1823118484,
              "labels": Array [],
              "location": Object {
                "column": 72,
                "line": 24,
              },
              "tag-name": "munit:run-custom",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteBefore_Test",
      "labels": Array [
        "munit",
        "vertical",
        "removed",
      ],
      "location": Object {
        "column": 91,
        "line": 22,
      },
      "tag-name": "munit:before-test",
      "type": "munit-container",
    },
    Object {
      "attributes": Object {
        "description": "MUnit Test",
        "name": "new-test-suiteTest",
      },
      "content": Array [
        Object {
          "content": null,
          "description": "Setup",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "doc:name": "Set Message",
                "payload": "#[]",
              },
              "content-hash": -2017569654,
              "description": "Set Message",
              "hash": -339828259,
              "labels": Array [],
              "location": Object {
                "column": 57,
                "line": 27,
              },
              "tag-name": "munit:set",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Fail",
              },
              "content-hash": -2017569654,
              "description": "Fail",
              "hash": 369682545,
              "labels": Array [],
              "location": Object {
                "column": 37,
                "line": 28,
              },
              "tag-name": "munit:fail",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Set Null Payload",
              },
              "content-hash": -2017569654,
              "description": "Set Null Payload",
              "hash": -852187714,
              "labels": Array [],
              "location": Object {
                "column": 61,
                "line": 29,
              },
              "tag-name": "munit:set-null-payload",
              "type": "component",
            },
          ],
          "description": "Test",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteTest",
      "labels": Array [
        "munit",
        "vertical",
        "removed",
      ],
      "location": Object {
        "column": 67,
        "line": 26,
      },
      "tag-name": "munit:test",
      "type": "munit-container",
    },
  ],
  "description": null,
  "hash": -2009730132,
  "labels": Array [
    "vertical",
    "root",
  ],
  "location": Object {
    "column": 108,
    "line": 12,
  },
  "tag-name": "mule",
  "type": "container",
}
`;

exports[`when diffing two XML files when doing a diff with a non existant initial state augments the MAST correctly 1`] = `
Object {
  "attributes": Object {},
  "content": Array [
    Object {
      "content": null,
      "content-hash": -2017569654,
      "description": "MUnit configuration",
      "doc:name": "MUnit configuration",
      "hash": -910446770,
      "labels": Array [
        "added",
      ],
      "location": Object {
        "column": 63,
        "line": 13,
      },
      "name": "munit",
      "tag-name": "munit:config",
      "type": "component",
    },
    Object {
      "content": Array [
        Object {
          "content": null,
          "description": "After suite",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "condition": "#[]",
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert False",
              "doc:name": "Assert False",
              "hash": -964144491,
              "labels": Array [],
              "location": Object {
                "column": 69,
                "line": 15,
              },
              "tag-name": "munit:assert-false",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteAfter_Suite",
      "labels": Array [
        "munit",
        "vertical",
        "added",
      ],
      "location": Object {
        "column": 89,
        "line": 14,
      },
      "name": "new-test-suiteAfter_Suite",
      "tag-name": "munit:after-suite",
      "type": "munit-container",
    },
    Object {
      "content": Array [
        Object {
          "content": null,
          "description": "After test",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "actualValue": "#[]",
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert Not Equals",
              "doc:name": "Assert Not Equals",
              "expectedValue": "#[]",
              "hash": 1131396067,
              "labels": Array [],
              "location": Object {
                "column": 99,
                "line": 18,
              },
              "tag-name": "munit:assert-not-same",
              "type": "component",
            },
            Object {
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert Not Null Payload",
              "doc:name": "Assert Not Null Payload",
              "hash": -802864801,
              "labels": Array [],
              "location": Object {
                "column": 67,
                "line": 19,
              },
              "tag-name": "munit:assert-not-null",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteAfter_Test",
      "labels": Array [
        "munit",
        "vertical",
        "added",
      ],
      "location": Object {
        "column": 88,
        "line": 17,
      },
      "name": "new-test-suiteAfter_Test",
      "tag-name": "munit:after-test",
      "type": "munit-container",
    },
    Object {
      "content": Array [
        Object {
          "content": null,
          "description": "Before test",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "condition": "#[]",
              "content": null,
              "content-hash": -2017569654,
              "description": "Assert True",
              "doc:name": "Assert True",
              "hash": 1099203488,
              "labels": Array [],
              "location": Object {
                "column": 67,
                "line": 22,
              },
              "tag-name": "munit:assert-true",
              "type": "component",
            },
            Object {
              "assertion-ref": "",
              "content": null,
              "content-hash": -2017569654,
              "description": "Custom Assertion",
              "doc:name": "Custom Assertion",
              "hash": 1823118484,
              "labels": Array [],
              "location": Object {
                "column": 72,
                "line": 23,
              },
              "tag-name": "munit:run-custom",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteBefore_Test",
      "labels": Array [
        "munit",
        "vertical",
        "added",
      ],
      "location": Object {
        "column": 91,
        "line": 21,
      },
      "name": "new-test-suiteBefore_Test",
      "tag-name": "munit:before-test",
      "type": "munit-container",
    },
    Object {
      "content": Array [
        Object {
          "content": null,
          "description": "Setup",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "content": null,
              "content-hash": -2017569654,
              "description": "Set Message",
              "doc:name": "Set Message",
              "hash": -339828259,
              "labels": Array [],
              "location": Object {
                "column": 57,
                "line": 26,
              },
              "payload": "#[]",
              "tag-name": "munit:set",
              "type": "component",
            },
            Object {
              "content": null,
              "content-hash": -2017569654,
              "description": "Set Null Payload",
              "doc:name": "Set Null Payload",
              "hash": -852187714,
              "labels": Array [],
              "location": Object {
                "column": 61,
                "line": 27,
              },
              "tag-name": "munit:set-null-payload",
              "type": "component",
            },
          ],
          "description": "Test",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "location": null,
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "new-test-suiteTest",
      "labels": Array [
        "munit",
        "vertical",
        "added",
      ],
      "location": Object {
        "column": 67,
        "line": 25,
      },
      "name": "new-test-suiteTest",
      "tag-name": "munit:test",
      "type": "munit-container",
    },
  ],
  "description": null,
  "hash": -15128758,
  "labels": Array [
    "vertical",
    "root",
  ],
  "location": Object {
    "column": 44,
    "line": 0,
  },
  "tag-name": "mule",
  "type": "container",
}
`;

exports[`when diffing two XML files when file B adds elements with nested content augments the MAST correctly 1`] = `
Object {
  "attributes": Object {
    "xmlns": "http://www.mulesoft.org/schema/mule/core",
    "xmlns:amqps": "http://www.mulesoft.org/schema/mule/amqps",
    "xmlns:doc": "http://www.mulesoft.org/schema/mule/documentation",
    "xmlns:dw": "http://www.mulesoft.org/schema/mule/ee/dw",
    "xmlns:objectstore": "http://www.mulesoft.org/schema/mule/objectstore",
    "xmlns:spring": "http://www.springframework.org/schema/beans",
    "xmlns:tracking": "http://www.mulesoft.org/schema/mule/ee/tracking",
    "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
    "xsi:schemaLocation": "http://www.mulesoft.org/schema/mule/ee/dw http://www.mulesoft.org/schema/mule/ee/dw/current/dw.xsd
http://www.mulesoft.org/schema/mule/amqps http://www.mulesoft.org/schema/mule/amqps/current/mule-amqps.xsd
http://www.mulesoft.org/schema/mule/objectstore http://www.mulesoft.org/schema/mule/objectstore/current/mule-objectstore.xsd
http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd
http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/ee/tracking http://www.mulesoft.org/schema/mule/ee/tracking/current/mule-tracking-ee.xsd",
  },
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": null,
          "content-hash": 1725153786,
          "description": "Some transform",
          "doc:name": "Some transform",
          "hash": 1005263453,
          "labels": Array [],
          "location": Object {
            "column": 56,
            "line": 16,
          },
          "tag-name": "dw:transform-message",
          "type": "component",
        },
      ],
      "description": "a sub flow",
      "hash": 1871898525,
      "labels": Array [
        "horizontal",
        "added",
      ],
      "location": Object {
        "column": 32,
        "line": 15,
      },
      "name": "a sub flow",
      "tag-name": "sub-flow",
      "type": "container",
    },
  ],
  "description": null,
  "hash": -1827547211,
  "labels": Array [
    "vertical",
    "root",
  ],
  "location": Object {
    "column": 126,
    "line": 13,
  },
  "tag-name": "mule",
  "type": "container",
}
`;
